### PR TITLE
feat: add request for data packages with biggest deviation in sdk

### PR DESCRIPTION
**Pull request by cehali**

### CODE REVIEW

File: packages/sdk/test/mocks/mock-packages.ts [hunk 1]
```diff
@@ -25,6 +25,32 @@ const signedDataPackagesObjects = [
...
+      { dataFeedId: "BTC", value: 20000 },
+      { dataFeedId: "ETH", value: 1002 },
+    ],
+    timestampMilliseconds: 1654353400000,
+    signature:
+      "NX5yd/Cs8HzVdNchrM59uOoSst7n9KK5Ou9pA6S5GTM0RwghGlFjA0S+SVfb85ipg4HzUTKATBZSqPXlWldEEhw=",
+    dataServiceId: "service-1",
+    dataFeedId: "ETH",
+    sources: null,
+    signerAddress: "0x2",
```
Comment (koonopek) [inline packages/sdk/test/mocks/mock-packages.ts:52]: Maybe we could change it to `0x3` and then assert in test (to better imitate real situation)

**Review by koonopek — APPROVED** [review]
GJ! 
One small comment you can ignore it

File: packages/on-chain-relayer/src/run-relayer.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { getAdapterContract } from "./core/contract-interactions/get-contract";
 import { getValuesForDataFeeds } from "./core/contract-interactions/get-values-for-data-feeds";
 import { sendHealthcheckPing } from "./core/monitoring/send-healthcheck-ping";
 import { config } from "./config";
+import { ValuesForDataFeeds } from "./types";
```
Comment (hatskier) [inline packages/on-chain-relayer/src/run-relayer.ts:10]: Shouldn't we have this type in redstone-sdk?

File: packages/sdk/src/index.ts [hunk 3]
```diff
@@ -132,6 +154,34 @@ export const getUrlsForDataServiceId = (
...
+  return dataPackages.sort((leftDataPackage, rightDataPackage) => {
+    const leftValue = Number(leftDataPackage.dataPoints[0].value);
+    const leftValueDeviation = calculateDeviation(leftValue, valueToCompare);
+    const rightValue = Number(rightDataPackage.dataPoints[0].value);
+    const rightValueDeviation = calculateDeviation(rightValue, valueToCompare);
+    return rightValueDeviation - leftValueDeviation;
+  });
+};
+
+const calculateDeviation = (value: number, valueToCompare: number) => {
```
Comment (hatskier) [inline packages/sdk/src/index.ts:177]: We repeat this function in at least 3 places now. We definitely need to move it to a common module in monorepo. But let's do it in a separate task

File: packages/sdk/src/index.ts [hunk 4]
```diff
@@ -132,6 +154,34 @@ export const getUrlsForDataServiceId = (
...
 };
 
+const sortDataPackagesByDeviationDesc = (
+  dataPackages: SignedDataPackagePlainObj[],
+  valueToCompare: number,
+  dataFeedsIds: string[]
+) => {
+  if (dataFeedsIds[0] === ALL_FEEDS_KEY) {
+    throw new Error(
+      `Cannot get data packages with biggest deviation for ${ALL_FEEDS_KEY}`
```
Comment (hatskier) [inline packages/sdk/src/index.ts:164]: `Cannot sort datapackages by deviation for ${ALL_FEEDS_KEY}`?

File: packages/sdk/src/index.ts [hunk 5]
```diff
@@ -132,6 +154,34 @@ export const getUrlsForDataServiceId = (
   return resolveDataServiceUrls(reqParams.dataServiceId);
 };
 
+const sortDataPackagesByDeviationDesc = (
+  dataPackages: SignedDataPackagePlainObj[],
+  valueToCompare: number,
+  dataFeedsIds: string[]
```
Comment (hatskier) [inline packages/sdk/src/index.ts:160]: Can we move this check to run it only one time and also check `dataFeedsIds.includes(ALL_FEEDS_KEY)`?

File: packages/sdk/src/index.ts [hunk 6]
```diff
@@ -66,8 +70,26 @@ export const parseDataPackagesResponse = (
       );
     }
 
-    parsedResponse[dataFeedId] = dataFeedPackages
-      .sort((a, b) => b.timestampMilliseconds - a.timestampMilliseconds) // we prefer newer data packages in the first order
+    let dataFeedPackagesSorted = dataFeedPackages;
+    if (
+      reqParams.valuesToCompare &&
+      Object.keys(reqParams.valuesToCompare).length > 0
```
Comment (hatskier) [inline packages/sdk/src/index.ts:76]: We can remove this check

File: packages/sdk/src/index.ts [hunk 7]
```diff
@@ -66,8 +70,26 @@ export const parseDataPackagesResponse = (
...
+    ) {
+      const decimals =
+        (dataFeedPackages[0].dataPoints[0] as INumericDataPoint).decimals ??
+        DEFAULT_DECIMALS;
+      const valueToCompare = Number(
```
Comment (hatskier) [inline packages/sdk/src/index.ts:81]: Let's check if it is not undefined `reqParams.valuesToCompare[dataFeedId]`, if it is we can skip sorting

File: packages/sdk/src/index.ts [hunk 8]
```diff
@@ -66,8 +70,26 @@ export const parseDataPackagesResponse = (
...
 
-    parsedResponse[dataFeedId] = dataFeedPackages
-      .sort((a, b) => b.timestampMilliseconds - a.timestampMilliseconds) // we prefer newer data packages in the first order
+    let dataFeedPackagesSorted = dataFeedPackages;
+    if (
+      reqParams.valuesToCompare &&
+      Object.keys(reqParams.valuesToCompare).length > 0
+    ) {
+      const decimals =
+        (dataFeedPackages[0].dataPoints[0] as INumericDataPoint).decimals ??
```
Comment (hatskier) [inline packages/sdk/src/index.ts:79]: Let's try to improve this logic and move it to protocol. Let's add there a method like `getDecimalsForDataFeedId(dataPackages: SignedDataPackages[], dataFeedId: string)`

File: packages/sdk/src/index.ts [hunk 9]
```diff
@@ -1,27 +1,35 @@
...
 const ALL_FEEDS_KEY = "___ALL_FEEDS___";
+const DEFAULT_DECIMALS = 8;
 
 export interface DataPackagesRequestParams {
   dataServiceId: string;
   uniqueSignersCount: number;
   dataFeeds?: string[];
   disablePayloadsDryRun?: boolean;
   urls?: string[];
+  valuesToCompare?: { [dataFeedId: string]: BigNumber };
```
Comment (hatskier) [inline packages/sdk/src/index.ts:22]: Maybe `valuesToCompare?: ValuesForDataFeeds`?

File: packages/sdk/src/index.ts [hunk 10]
```diff
@@ -77,6 +90,51 @@ export const parseDataPackagesResponse = (
...
+const getDataPackagesSortedByDeviation = (
+  dataFeedPackages: SignedDataPackagePlainObj[],
+  valuesToCompare: ValuesForDataFeeds | undefined,
+  dataFeedId: string
+) => {
+  if (!valuesToCompare) {
+    return dataFeedPackages;
+  }
+
+  if (dataFeedId === ALL_FEEDS_KEY) {
```
[the remark this answers is not in the document]
Comment (cehali) [reply packages/sdk/src/index.ts:102]: We use `dataFeedId` also for `valuesToCompare` to get proper value

**Review by hatskier — APPROVED** [review]

**Pull request merged**